### ISSUE TITLE
Skaffold sometimes doesn't deploy all services

### ISSUE THREAD

**Issue by ChaseMor** [GoogleContainerTools/skaffold#4519]
### Expected behavior
All services should deploy all the time

### Actual behavior
A fraction of the time, a service gets lost and doesn't deploy

### Information

- Skaffold version: version v1.12.0
- Operating system: Mac 10.15.5
- Contents of skaffold.yaml:

```yaml
apiVersion: skaffold/v1beta15
kind: Config
build:
  tagPolicy:
    sha256: {}
  # defines where to find the code at build time and where to push the resulting image
  artifacts:
  - image: dotnet-guestbook-backend
    context: src/backend
  - image: dotnet-guestbook-frontend
    context: src/frontend
# defines the Kubernetes manifests to deploy on each run
deploy:
  kubectl:
    manifests:
    - ./kubernetes-manifests/*.service.yaml
    - ./kubernetes-manifests/*.deployment.yaml
profiles:
# use the cloudbuild profile to build images using Google Cloud Build
- name: cloudbuild
  build:
    googleCloudBuild: {}
```

### Steps to reproduce the behavior

1. Use either the Cloud Code VS Code or IntelliJ extension
2. Create a new application using the Java guestbook sample
3. Select run on Kubernetes
4. Observe that about 5-10% of the time, the frontend service doesn't deploy

#### VS Code
![Screen Shot 2020-07-21 at 12 39 54 AM](https://user-images.githubusercontent.com/13285164/88106010-a5a80000-cb59-11ea-81f4-1429b74db723.png)

#### IntelliJ
![Screen Shot 2020-07-21 at 1 31 32 PM](https://user-images.githubusercontent.com/13285164/88106015-a771c380-cb59-11ea-81ab-8207c4f8b79a.png)

**Comment by sivakku** [comment]
@nkubala/@briandealwis - Can you triage and label it as july-chill? i can't edit Labels in this repository.

**Comment by dgageot** [comment]
I couldn't reproduce that behaviour after many many tries.
Are you deploying to minikube or Docker Desktop?

**Comment by dgageot** [comment]
The one hint that's visible from Skaffold's output in IntelliJ is that the `java-guestbook-frontend` was already deployed to the cluster before Skaffold deploys everything.
<img width="335" alt="Screenshot 2020-07-22 at 09 55 03" src="https://user-images.githubusercontent.com/153495/88150338-8cf32500-cc01-11ea-8d28-afbb4c38685d.png">

**Comment by dgageot** [comment]
Here's what I think happens (but I couldn't reproduce):
 + Skaffold deploys the whole application
 + The `java-guestbook-frontend` object already existed, with at least one difference. So it has to be re-configured
 + By the time Skaffold queries for all deployed services, the `java-guestbook-frontend` is not part of the results. Probably because it's in progress of applying the reconfiguration.
 + Skaffold doesn't start port forwarding for that service

I'll try to put a service in an unstable stable.

**Comment by dgageot** [comment]
Thank you @ChaseMor for sharing the procedure that you used to reproduce:

 + Using VSCode (or IntelliJ), "Run on Kubernetes", to a **GKE cluster**
 + Wait for the port-forwarding information to be displayed
 + Check if the `service/java-guestbook-frontend` is forwarded
 + Stop the app
 + Try again until `service/java-guestbook-frontend` is not forwarded

**Comment by dgageot** [comment]
OK, I couldn't really reproduce the same issue but I still see a strange behaviour with the following steps:

 + Using VSCode, I deploy to a GKE cluster with "Run on Kubernetes"
 + Wait for the port-forwarding information to be displayed
 + All the services `service/java-guestbook-frontend` are forwarded correctly
 + **I open the cloud console and wait until the service has an external IP allocated**
 + Stop the app
 + Deploy again the app
 + All the services `service/java-guestbook-frontend` are forwarded correctly
 + **I open the cloud console, wait for about 10s and refresh. The frontend service is gone. All other services that are not of type LoadBalancer are still there.**

So there's at least a race condition on the GKE side that deletes the service after Skaffold has created it the second time.

Here's the flow:
 + First run creates a LoadBalancer service with `kubectl apply`
 + When we stop the app, the service is deleted with `kubectl delete`
 + Second run creates a new LoadBalancer service with `kubectl apply`, This service has the exact same configuration as the first one except for a `skaffold.dev/run-id` label that is set to a unique ID by Skaffold.
 + After a few seconds, this second service disappears on the GKE side

I couldn't reproduce the race condition on Docker Desktop probably because deleting a LoadBalancer service is much faster.

**Comment by dgageot** [comment]
I think what's happening is that the deletion phase of the first run is not yet fully done that the user runs a new deploy.

Because deleting a LoadBalancer service on GKE takes many seconds, half way through the second deploy, the deletion completes.

If it completes before Skaffold sets up port forwarding, no port forwarding message is shown for that service.

Now, how can we fix that?
 + It's not clear to the user that Skaffold has actually deleted all the resources. It show `deployment.apps "java-guestbook-frontend" delete` but then waits for it to be really deleted and only then does Skaffold's process end. If it was clearer that the deletion was in progress, maybe the user wouldn't redeploy their app.
 + VSCode and IntelliJ could wait for the previous deploy to actually terminate before they start a new one.
 + We could investigate more if there's a race condition on the GKE side. I wonder what the k8s spec says about deleting a service and recreating it before it's fully deleted.

**Comment by briandealwis** [comment]
Nice analysis.  Your flow makes sense: the service's name, type, selector, and ports are all the same — `apply` would just apply the label changes and return.

`kubectl delete`'s output is very sparse.  A status-check-like live-reporting of deletion status might be helpful.  I know I've sometimes had deletes sometimes take what feels like minutes, and I've ^C to get a move on.

**Comment by sivakku** [comment]
Thanks @dgageot for analysis. 

- Waiting for delete of older instance from user perspective is a little bit strained experience. This makes sense on production/staging but on minikube we might want rapid experience.
- For the newer run-id, why does it look at the older instance even if it has same configuration and decides the newer configuration based on it? We can look at GKE to reduce time to cleanup but i think there will be always similar issues that time outs might not be appropriate for minikube.

**Comment by dgageot** [comment]
@sivakku On minikube, the problem will not really happen. Deleting a LoadBalancer service is very quick. That's not the case on GKE.

**Comment by nkubala** [comment]
nice track down @dgageot. I think your fix makes sense, but I'm wondering if we need to wait for all resource deletions, or if the problem is really only with ones that take a long time to delete, i.e. LoadBalancer/Ingress. I doubt we would see this issue with pods/deployments

**Comment by briandealwis** [comment]
Note that a pod's termination grace period is normally 30 seconds, but can be set to an arbitrary integer, so it could be a while (and [potentially forever!](https://github.com/kubernetes/kubernetes/issues/84298)).  Note that this should only wait on objects marked as being deleted.  So the downsides is that this will introduce 1 more kubectl call  when there are no objects being delete.

**Comment by dgageot** [comment]
@briandealwis I think the downside is ok. Don't you think?

**Comment by sivakku** [comment]
Future dependent work items on Kubectl from David, to make it more paralleized.

- Detect apply on deleting resources: https://github.com/kubernetes/kubernetes/issues/93363
- Allow foreground deletion in kubectl: https://github.com/kubernetes/kubernetes/issues/93364
- kubectl wait --for=delete fails on 404: https://github.com/kubernetes/kubernetes/issues/93365
- A quick PR to clarify the type of cascading on kubectl delete --cascade: https://github.com/kubernetes/kubernetes/pull/93367
